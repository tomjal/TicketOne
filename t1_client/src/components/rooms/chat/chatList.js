import React, { Component } from 'react';
import { css } from 'aphrodite';

import { chatStyles } from './../../../styles/inline/specificStyles';
import { timestampToDateString } from './../../../helpers/timeUtils';
import ChatBubble from './chatBubble';

export class ChatList extends Component {
  componentDidMount() {
    this.scrollToBottom();
  }
  componentDidUpdate() {
    this.scrollToBottom();
  }
  scrollToBottom() {
    if (this.messageList) {
      const scrollHeight = this.messageList.scrollHeight;
      const height = this.messageList.clientHeight;
      const maxScrollTop = scrollHeight - height;
      this.messageList.scrollTop = maxScrollTop > 0 ? maxScrollTop : 0;
    }
  }
  isMyBubble(role) {
    if (role === this.props.role) {
      return true;
    }
    return false;
  }
  render() {
    const { messages } = this.props;

    return (
      <div className={`${css(chatStyles.list)} room-chat-body`} ref={(div) => { this.messageList = div; }}>
        {messages &&
          <div >
            {
              messages.map((message, i) => {
                const isMe = this.isMyBubble(message.author.role);
                const dateString = timestampToDateString(message.timestamp);
                return <ChatBubble key={i} isMe={isMe} message={message} date={dateString} />
              })
            }
          </div>
        }
      </div>
    );
  }
}

export default ChatList;
