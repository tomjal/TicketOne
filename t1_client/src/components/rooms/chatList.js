import React, { Component } from 'react';

/*
message schema
{ body: "", author: { id: "", role: "senderRole" }, timestamp: "" }
*/

class ChatList extends Component {
  componentDidMount() {
    this.scrollToBottom();
  }

  componentDidUpdate() {
    this.scrollToBottom();
  }
  scrollToBottom() {
    const scrollHeight = this.messageList.scrollHeight;
    const height = this.messageList.clientHeight;
    const maxScrollTop = scrollHeight - height;
    this.messageList.scrollTop = maxScrollTop > 0 ? maxScrollTop : 0;
  }
  timestampToDateString(ts) {
    function pad(n) { return n < 10 ? '0' + n : n }

    let d = new Date(ts);
    let dateString = d.getDate() + '/' + (d.getMonth() + 1) + '/' + d.getFullYear();
    let timeString = pad(d.getHours()) + ':' + pad(d.getMinutes());
    return timeString + ' ' + dateString;
  }
  isMyBubble(role) {
    if (role == this.props.role) {
      return true;
    }
    return false;
  }
  render() {
    const { messages } = this.props;
    const chatListStyle = {
      width: "100%", height: "255px", overflow: "auto",
      padding: "10px", marginTop: "5px", marginBottom: "15px",
      background: "#f1f3f4", boxShadow: "0 1px 5px rgba(0,0,0,0.12), 0 1px 5px 0px rgba(0,0,0,0.24)"
    }

    const baseBubbleStyle = { width: "300px", background: "white", borderRadius: "5px", marginBottom: "10px", padding: "10px" };
    const leftBubbleStyle = { textAlign: "left" };
    const rightBubbleStyle = { textAlign: "right", marginLeft: "115px" };

    const headerMessageStyle = { color: "gray", fontSize: "1.2rem" };
    const bodyMessageStyle = { color: "black", fontSize: "1.4rem", wordWrap: "break-word" };

    let me = this;
    return (
      <div style={chatListStyle} ref={(div) => { this.messageList = div; }}>
        {messages &&
          <div >
            {
              messages.map(function (message, i) {
                let isMe = me.isMyBubble(message.author.role);
                let dateString = me.timestampToDateString(message.timestamp);
                let bubbleStyle = (isMe === true) ? { ...baseBubbleStyle, ...rightBubbleStyle } : { ...baseBubbleStyle, ...leftBubbleStyle };
                return <div style={bubbleStyle} key={i}>
                  <div style={headerMessageStyle}>{!isMe&&<span>{message.author.id} - </span>}{dateString}:</div>
                  <div style={bodyMessageStyle}>{message.body}</div>
                </div>;
              })
            }
          </div>
        }
      </div>
    );
  }
}

export default ChatList;
