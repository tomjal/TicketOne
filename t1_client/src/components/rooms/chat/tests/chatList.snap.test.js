import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { ChatList } from './../chatList';

test('ChatList snapshot', () => {
  const wrap = shallow(<ChatList />);

  expect(shallowToJson(wrap)).toMatchSnapshot();
});