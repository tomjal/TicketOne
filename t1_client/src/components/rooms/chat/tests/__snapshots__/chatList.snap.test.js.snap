exports[`test ChatList snapshot 1`] = `
<div
  className="list_y16a99 room-chat-body" />
`;
